src/main/java/com/android/volley/AsyncCache.java

```diff
@@ -0,0 +1,82 @@
+package com.android.volley;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicReference;
+
+public abstract class AsyncCache implements Cache {
+
+    public interface OnGetCompleteCallback {
+        void onGetComplete(Entry entry);
+    }
+
+    public abstract void get(String key, OnGetCompleteCallback callback);
+
+    // TODO: Handle callback never being invoked? Match behavior when sync cache throws
+    // RuntimeException.
+
+    @Override
+    public Entry get(String key) {
+        final CountDownLatch latch = new CountDownLatch(1);
+        final AtomicReference<Entry> entryRef = new AtomicReference<>();
+        get(
+                key,
+                new OnGetCompleteCallback() {
```

Comment:
nit: we could use a lambda expression here for simplicity (I believe [all version of Android can now use a lambda](https://developer.android.com/studio/write/java8-support) when using a recent Android Gradle plugin)

This would convert your code into:

```java
get(
    key, 
    entry -> {
      entryRef.set(entry);
      latch.countDown();
    }
);
```

For context, lambdas are like a more concise way to write an anonymous class. Feel free to take a look at the official java doc about lambdas if it's the first time you hear about them / use them: https://docs.oracle.com/javase/tutorial/java/javaOO/lambdaexpressions.html

Reply:
When I tried to make that change, it says lambda expressions aren't allowed in Java 7. Even though I thought Volley uses java8. I'll look into this.

Reply:
Volley actually is built with minSdkVersion = 8, which is actually Java 6 :)

Lambdas have been backported in principle per https://developer.android.com/studio/write/java8-support - not sure offhand how far back this goes. But I'm not 100% sure what the story is offhand in terms of exposing a library that relies on this support.

While it's not ideal, unless we can pretty quickly and confidently find an answer suggesting we can safely use lambdas without introducing major complexity for apps depending on the library using this desugaring support, we should probably avoid them except when writing code that is guaranteed to run on the right minSdkVersion.

Note that this also includes try-with-resources.

Reply:
FWIW, I'm open to considering increasing the minSdkVersion somewhat. At this point it's exceedingly rare to target anything under API 14, which is what the Android support library requires; that seems like a reasonable point to increase it, if that'd be helpful. But it'd be nice to have a clear benefit to doing so, and 14 won't get us native try-with-resources or lambdas AFAIK.

Reply:
To be clear, the current Java target will support try-with-resources if sdk >= 19 

```
compileOptions {
    sourceCompatibility JavaVersion.VERSION_1_7
    targetCompatibility JavaVersion.VERSION_1_7
}
```

Though I think try-with-resources can obscure what's going on, esp. in nested and other complex situations, so no loss of any significance.

https://stackoverflow.com/a/19470405/901597

Java 8 is needed for lambda.